File: data-structures-and-algorithms/linked-list/easy/reverse-linked-list/reverse_linked_list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        return solution(head)

def solution(head):
    newHead = None
    while head:
        tmp = head
        head = head.next
        tmp.next = newHead
        newHead = tmp
        
    return newHead

File: data-structures-and-algorithms/linked-list/easy/reverse-linked-list/reverseLinkedList.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
function reverseList(head) {
  let newHead = null

  // head is the current node in the linked list when we call `this.val`
  while (head != null) {
    let curr = head
    head = head.next
    curr.next = newHead
    newHead = curr
  }

  return newHead
}